lib/paramsx.ex

```diff
@@ -1,35 +1,85 @@
 defmodule Paramsx do
-  alias Paramsx.{Atomize, Filter}
-
   @moduledoc """
-    Documentation for Paramsx.
+    Paramsx provides functionally to whitelist and validate parameters
   """
 
   @doc """
-  Filter params based in your  required and optional keyword.
+  Filter params based on your required and optional keyword. 
 
   ## Examples
-    
       iex> Paramsx.filter(%{"foo" => "bar", "foo2" => "bar2"}, required: [:foo])
-      %{"foo" => "bar"}
+      {:ok, %{foo: "bar"}}
     
       iex> Paramsx.filter(%{"foo" => "bar", "foo2" => "bar2"}, required: [:foo3])
       {:error, %{missing_keys: [:foo3]}}
 
       iex> Paramsx.filter(%{"foo" => "bar", "foo2" => "bar2"}, required: [:foo], optional: [:foo3])
-      %{"foo" => "bar"}  
+      {:ok, %{foo: "bar"}}
+
+  ### Dont allow params list if it's not specified authorize only string and number
+     iex> Paramsx.filter(%{"foo" => %{"bar" => "value_bar"}}, required: [:foo])
+     {:ok, %{}}
+
+     iex> Paramsx.filter(%{"foo" => %{"bar" => "value_bar"}}, required: [foo: [:bar]])
+     {:ok, %{foo: %{bar: "value_bar"}}}
   """
 
-  def filter(params, filters), do: Filter.validate_for_presence(params, filters)
+  def filter(params, filters) when is_map(params) and is_list(filters) do
+    required = Keyword.get(filters, :required, [])
+    optional = Keyword.get(filters, :optional, [])
 
-  @doc """
-  Atomize keys values.
+    params
+    |> filter_required(required)
+    |> filter_optional(params, optional)
+  end
 
-  ## Examples
-    
-      iex> Paramsx.atomize_keys(%{"foo" => "bar"})
-      %{foo: "bar"}
-  """
+  defp filter_required(params, filters),
+    do: reduce_filters(filters, %{}, params, :required)
+
+  defp filter_optional(%{} = acc, params, filters),
+    do: {:ok, reduce_filters(filters, acc, params, :optional)}
+
+  defp filter_optional(missing, _params, _filters), do: {:error, %{missing_keys: missing}}
+
+  defp reduce_filters(filters, acc, params, mode),
+    do: Enum.reduce(filters, acc, &reduce_fun(&1, &2, params, mode))
+
+  defp reduce_fun([{key, filters}], acc, params, mode) when is_list(filters),
+    do: reduce_fun_for_nested(key, filters, acc, params, mode)
+
+  defp reduce_fun({key, filters}, acc, params, mode) when is_list(filters),
+    do: reduce_fun_for_nested(key, filters, acc, params, mode)
+
+  defp reduce_fun(key, %{} = acc, params, mode) do
+    case fetch(params, key) do
+      {:ok, value} when is_binary(value) or is_number(value) -> put(acc, key, value)
+      {:ok, _value} -> acc
+      _not_found -> handle_missing_key(mode, acc, key)
+    end
+  end
+
+  defp reduce_fun(key, missing, params, _mode) when is_list(missing),
```

Comment:
```elixir
defp reduce_fun(key, missing, params, _mode) when is_list(missing) do
  if Map.has_key?(params, to_string(key)) do
    missing
  else
    [key | missing]
  end
end
```
:nail_care: IMO Seems that we are forcing to be inline, the code seems a little bit strange.